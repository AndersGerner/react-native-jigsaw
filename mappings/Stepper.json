[{"name":"Stepper","tag":"Stepper","description":"A component used to control the quantity of something","category":"field","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Control_Stepper.png","supports_list_render":false,"props":{"fieldName":{"label":"Field name","description":"The name of the field within the screen that will store this component's value","type":"fieldName","value":"stepperValue","valuePropName":"value","handlerPropName":"onChange","editable":true,"required":false}},"layout":{}}]