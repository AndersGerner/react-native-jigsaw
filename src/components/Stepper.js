/* @flow */

import * as React from "react";
import { View, Image, Text } from "react-native";
import { withTheme } from "../core/theming";
import { COMPONENT_TYPES, FORM_TYPES } from "../core/component-types";
import IconButton from "./IconButton";
import Config from "./Config";
import type { Theme } from "../types";

type Props = {
  value: number,
  onPlus: () => void,
  onMinus: () => void,
  style?: any,
  theme: Theme
};

class Stepper extends React.Component<Props> {
  handleMinus = () => {
    const { value, onMinus } = this.props;

    if (value > 0) {
      onMinus();
    }
  };

  render() {
    const {
      value,
      onPlus,
      theme: { colors, spacing, typography },
      style
    } = this.props;

    return (
      <View style={[{ flexDirection: "row" }, style]}>
        <IconButton
          icon="remove"
          onPress={this.handleMinus}
          size={Config.stepperButtonSize}
          color={colors.strong}
          disabled={value === 0}
        />
        <Text
          style={[
            typography.body1,
            {
              textAlign: "center",
              alignSelf: "center",
              color: colors.medium,
              marginHorizontal: spacing.medium
            }
          ]}
        >
          {value}
        </Text>
        <IconButton
          icon="add"
          onPress={onPlus}
          size={Config.stepperButtonSize}
          color={colors.strong}
        />
      </View>
    );
  }
}

export default withTheme(Stepper);

export const SEED_DATA = [
  {
    name: "Stepper",
    tag: "Stepper",
    description: "A component used to control the quantity of something",
    category: COMPONENT_TYPES.field,
    preview_image_url:
      "https://res.cloudinary.com/altos/image/upload/v1541096671/draftbit/library/jigsaw-1.0/reps/Control_Stepper.png",
    supports_list_render: false,
    props: {
      value: {
        label: "Value",
        description: "Current value of stepper",
        type: FORM_TYPES.number,
        value: 0,
        min: 0,
        max: 999,
        step: 1,
        precision: 0,
        editable: true
      },
      onMinus: {
        label: "Stepper onMinus Function",
        description: "Function to run when minus button pressed",
        editable: true,
        type: FORM_TYPES.function,
        value: "{this.onMinus}"
      },
      onPlus: {
        label: "Stepper onPlus Function",
        description: "Function to run when minus button pressed",
        editable: true,
        type: FORM_TYPES.function,
        value: "{this.onPlus}"
      }
    },
    layout: {
      width: 126,
      height: 42
    }
  }
];
