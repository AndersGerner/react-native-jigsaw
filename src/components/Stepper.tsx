import * as React from "react";
import { View, Text, StyleProp, ViewStyle } from "react-native";
import { withTheme } from "../core/theming";
import { COMPONENT_TYPES, FIELD_NAME } from "../core/component-types";
import IconButton from "./IconButton";
import Config from "./Config";
import theme from "../styles/DefaultTheme";

interface Props {
  value?: number;
  theme: typeof theme;
  style?: StyleProp<ViewStyle>;
  onChange?: (value: number) => void;
}

const Stepper: React.FC<Props> = ({
  value = 0,
  style,
  onChange,
  theme: { colors, typography, spacing },
}) => {
  const [stateValue, setStateValue] = React.useState(value);

  const handleMinus = () => {
    if (value || value === 0) {
      onChange && onChange(value - 1);
    } else {
      setStateValue(stateValue - 1);
    }
  };

  const handlePlus = () => {
    if (value || value === 0) {
      onChange && onChange(value + 1);
    } else {
      setStateValue(stateValue + 1);
    }
  };

  return (
    <View style={[{ flexDirection: "row" }, style]}>
      <IconButton
        icon="MaterialIcons/remove"
        onPress={handleMinus}
        size={Config.stepperButtonSize}
        color={colors.strong}
        disabled={value ? value === 0 : stateValue === 0}
      />
      <Text
        style={[
          typography.body1,
          {
            textAlign: "center",
            alignSelf: "center",
            color: colors.medium,
            marginHorizontal: spacing.medium,
          },
        ]}
      >
        {value || stateValue}
      </Text>
      <IconButton
        icon="MaterialIcons/add"
        onPress={handlePlus}
        size={Config.stepperButtonSize}
        color={colors.strong}
      />
    </View>
  );
};

export default withTheme(Stepper);

export const SEED_DATA = [
  {
    name: "Stepper",
    tag: "Stepper",
    description: "A component used to control the quantity of something",
    category: COMPONENT_TYPES.field,
    preview_image_url: "{CLOUDINARY_URL}/Control_Stepper.png",
    supports_list_render: false,
    props: {
      fieldName: {
        ...FIELD_NAME,
        defaultValue: "stepperValue",
      },
    },
    layout: {},
  },
];
