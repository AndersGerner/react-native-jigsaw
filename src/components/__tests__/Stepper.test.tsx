import * as React from "react";
import renderer from "react-test-renderer";
import Stepper from "../Stepper";

it("renders stepper", () => {
  const tree = renderer.create(<Stepper />).toJSON();

  expect(tree).toMatchSnapshot();
});
