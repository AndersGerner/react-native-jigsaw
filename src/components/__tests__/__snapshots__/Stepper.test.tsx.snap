// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders stepper 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
      },
      undefined,
      Object {
        "borderRadius": 24,
      },
    ]
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 0.5,
      }
    }
  >
    <View>
      <View />
    </View>
  </View>
  <Text
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "letterSpacing": 0,
          "lineHeight": 26,
        },
        Object {
          "alignSelf": "center",
          "color": "rgba(70, 78, 88, 1)",
          "marginHorizontal": 8,
          "textAlign": "center",
        },
        undefined,
      ]
    }
  >
    0
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    hitSlop={
      Object {
        "bottom": 6,
        "left": 6,
        "right": 6,
        "top": 6,
      }
    }
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "opacity": 1,
      }
    }
  >
    <View>
      <View />
    </View>
  </View>
</View>
`;
